8336843: Deprecate java.util.zip.ZipError for removal

eirbjo opened the pull request:
Please review this PR which suggests to deprecate the unused class `java.util.zip.ZipError` for removal.

The class has been unsed by OpenJDK since  the ZIP API was rewritten from native to Java in JDK 9.

I opted to not explain the reason for the deprecation in detail, but instead simply point to `ZipException` as an alternative. Should more explanation be desired, I could prepend that with a note saying that the class is unused since JDK 9.

A CSR for this API update has been drafted, I'll update the Specification section there once we reach a concensus on the deprecation note in this PR.

This deprecation was initially suggested here: https://mail.openjdk.org/pipermail/core-libs-dev/2024-June/125720.html

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8338663](https://bugs.openjdk.org/browse/JDK-8338663) to be approved
- [x] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issues
 * [JDK-8336843](https://bugs.openjdk.org/browse/JDK-8336843): Deprecate java.util.zip.ZipError for removal (**Enhancement** - P4)
 * [JDK-8338663](https://bugs.openjdk.org/browse/JDK-8338663): Deprecate java.util.zip.ZipError for removal (**CSR**)


### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20642/head:pull/20642` \
`$ git checkout pull/20642`

Update a local copy of the PR: \
`$ git checkout pull/20642` \
`$ git pull https://git.openjdk.org/jdk.git pull/20642/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20642`

View PR using the GUI difftool: \
`$ git pr show -t 20642`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20642.diff">https://git.openjdk.org/jdk/pull/20642.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20642#issuecomment-2307283941)

eirbjo commented:
/csr

LanceAndersen commented:
On a corpus search, found
org.eclipse.jdt.core/search/org/eclipse/jdt/internal/core/search/indexing/AddJarFileToIndex.java references ZipError (16 usages)

sbt/internal/inc/zip/ZipCentralDir.java referenced ZipError(was based off an old copy of ZipFS I believe), the file does not seem to be part of the GitHub repository and there are 5 usages

com/aoapps/lang/Throwables.class  and com/aoindustries/lang/Throwables.class (clone of each other I believe) 

com/android/tools/apk/analyzer/internal/ArchiveManagerImpl. ( 1 usage)


So would be good to also alert those projects if possible

ExE-Boss commented:
There’s also [HMCL’s `org.jackhuang.hmcl.util.io.CompressingUtils`][1], which catches `ZipError` to convert it to a `ZipException` when run on **Java 8** ([this was also mentioned in the linked core‑libs‑dev discussion][2]).

[1]: https://github.com/HMCL-dev/HMCL/blob/95a1496389e6a8c2f51697706a20055a7952b8a5/HMCLCore/src/main/java/org/jackhuang/hmcl/util/io/CompressingUtils.java#L218-L220
[2]: https://mail.openjdk.org/pipermail/core-libs-dev/2024-June/125724.html

LanceAndersen commented:
> There’s also [HMCL’s `org.jackhuang.hmcl.util.io.CompressingUtils`](https://github.com/HMCL-dev/HMCL/blob/95a1496389e6a8c2f51697706a20055a7952b8a5/HMCLCore/src/main/java/org/jackhuang/hmcl/util/io/CompressingUtils.java#L218-L220), which catches `ZipError` to convert it to a `ZipException` when run on **Java 8** ([this was also mentioned in the linked core‑libs‑dev discussion](https://mail.openjdk.org/pipermail/core-libs-dev/2024-June/125724.html)).

Yes that is true.  The difference there is that is due to the demo version of Zip FS and as of JDK 9, when Zip FS was supported, it did not throw ZipError

liach commented:
A search on GitHub shows multiple Minecraft-related Java projects catch (but not create) ZipError for Java 8 backward compatibility. Maybe the actual deprecation can take steps; first with a functional removal to make ZipError constructors and methods throw errors, before finally removing it many releases later.

eirbjo commented:
> So would be good to also alert those projects if possible

I guess the deprecation warning itself is the "official" channel for communicating this type of information. But I agree it doesn't hurt to alert at least some of the larger projects, on a best-effort bases. So I have reached out to the following projects by reporting issues:

* Eclipse JDT: https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2856
* SBT / Zinc: https://github.com/sbt/zinc/issues/1385    
* Android Studio apkparser: https://issuetracker.google.com/issues/361862739

liach commented:
@eirbjo I have updated the CSR to describe the solutions from both our and user sides.

I think you can update the class Javadoc to something like:

```java
/**
 * A legacy error for invalid ZIP format from argument incorrectly regarded as an internal error.
 * This extends {@code InternalError} to maintain compatibility with preexisting client code catching
 * {@code InternalError}.
 * 
 * @deprecated
 * This error is not suitable to indicate an exception for a general ZIP file; use {@link
 * ZipException} instead.
 * <p>
 * {@code ZipError} fell out of use in the reference implementation in release 9 when the native ZIP
 * API was brought over to Java. Its last uses was in ZipFile and the demo zipfs in release 8.
 * <p>
 * Code bases supporting release 8 should update to catch InternalError instead. Old binaries can
 * use bytecode patching to upgrade CONSTANT_Class references to {@code java/util/zip/ZipError} in
 * exception handler ranges to {@code java/lang/InternalError}.
 */
```

LanceAndersen commented:
> @eirbjo I have updated the CSR to describe the solutions from both our and user sides.
> 
> I think you can update the class Javadoc to something like:
> 
> ```java
> /**
>  * A legacy error for invalid ZIP format from argument incorrectly regarded as an internal error.
>  * This extends {@code InternalError} to maintain compatibility with preexisting client code catching
>  * {@code InternalError}.
>  * 
>  * @deprecated
>  * This error is not suitable to indicate an exception for a general ZIP file; use {@link
>  * ZipException} instead.
>  * <p>
>  * {@code ZipError} fell out of use in the reference implementation in release 9 when the native ZIP
>  * API was brought over to Java. Its last uses was in ZipFile and the demo zipfs in release 8.
>  * <p>
>  * Code bases supporting release 8 should update to catch InternalError instead. Old binaries can
>  * use bytecode patching to upgrade CONSTANT_Class references to {@code java/util/zip/ZipError} in
>  * exception handler ranges to {@code java/lang/InternalError}.
>  */
> ```
> 
> I don't know if we should include such an excerpt:
> 
> ```java
> /**
>  * <p>
>  * In a future release, implementation of {@code ZipError} will be removed before this class itself
>  * is removed; at that point, {@code ZipError} can no longer be constructed or used, but can still
>  * be compiled against as an aid in migration.
>  */
> ```
> 
> I left the questions in CSR comment section; hope Joe Darcy can give us some advice here.

While I agree an `@deprecated` note should be added, I do not feel it needs to be that verbose.  I would simply borrow from what was done for `SecurityManager`.  The CSR can provide some more background


I also do not see a need to anything WRT the constructors as that is overkill IMHO.   

A release note will also go with this change which can  suggest that code that also must support JDK 8 leverage InternalError

liach commented:
I think linking to the CSR would be fine, but there is no prerequisite for API specs to link to CSRs. Need @jddarcy to double check here. I was emulating `Thread.stop`: https://github.com/openjdk/jdk/blob/5671f836039ef1683e3e9ce5b7cf0fa2f1860e2d/src/java.base/share/classes/java/lang/Thread.java#L1637

LanceAndersen commented:
> I think linking to the CSR would be fine, but there is no prerequisite for API specs to link to CSRs. Need @jddarcy to double check here. I was emulating `Thread.stop`:
> 
> https://github.com/openjdk/jdk/blob/5671f836039ef1683e3e9ce5b7cf0fa2f1860e2d/src/java.base/share/classes/java/lang/Thread.java#L1637

I am not suggesting link to the CSR, what I was indicating the CSR provides more details  because of the proposed removal.

Comparing to `Thread::stop` is not quite the same, and  the verbiage you are pointing to above was added when the method was [degraded to throw a UOE](https://bugs.openjdk.org/browse/JDK-8277861), not  when it was [deprecated for removal](https://bugs.openjdk.org/browse/JDK-8277861)

The overall usage of ZipError is extremely minimal in the wild at best, from what I can tell was in a catch statement and was not documented to be thrown by any public  API, though was thrown by ZipFile's internal ZipEntryIterator::next

If this was a more heavily used Exception, it would probably warrant further consideration for additional documentation, I am not convinced it is needed here

eirbjo commented:
JEP-277, "Enhanced Deprecation" includes the following recommendations:

*It is strongly recommended that the reasons for deprecating an API be described in that API's documentation comments. In addition, it is also recommended that potential replacement APIs be discussed and linked from the documentation.*

The reason we are deprecating this class is mainly that it became obsolete in JDK 9. I think we should say just that, and not elaborate with more details of the history of the class.

The recommended API replacement depends a bit on whether the code catches or throws `ZipError`, but also depends on whether the code needs to support running on Java 8 and catching `ZipError` there. Here I think we should mention the use case of catching ZipError when running on Java 8 and briefly point in the direction of the super class `InternalError`.

I think I would prefer to not update the class comment.

Based on the above, the PR now suggests the following documentation comment:

```
 * Signals that an unrecoverable error has occurred.
 * @deprecated This error became obsolete in JDK 9. Use
 * {@link ZipException} instead. Code needing to catch this
 * error when running on JDK 8 or earlier releases may catch
 * the parent {@link InternalError} instead.
 */

liach reviewed (approved):
/reviewers 2 reviewer

The updated deprecation specification looks good to me. I will review the CSR when you update it. Needs Lance or another engineer in the area to confirm too.

eirbjo commented on src/java.base/share/classes/java/util/zip/ZipError.java:
```diff
@@ -27,7 +27,10 @@
+ * {@link ZipException} instead. Code needing to catch this
+ * error when running on JDK 8 or earlier releases may catch
+ * the parent {@link InternalError} instead.
  * @author  Dave Bristor
```
> The javadoc should be focused on the current JDK release, not the prior history of this Class

Makes sense, let's focus on the current release.

> ```
>  * @deprecated ZipError is deprecated and subject to removal in a
>  *  future release. This class is obsolete. Use  {@link ZipException} instead.
> ```

The leading sentence here seems redundant, since it mostly just repeats the heading generated by javadoc:

**Deprecated, for removal: This API element is subject to removal in a future version**
_ZipError is deprecated and subject to removal in a future release. This Error is obsolete and no longer thrown. Use ZipException instead._

What do you think of dropping that first sentence, such that the rendered result would be:

**Deprecated, for removal: This API element is subject to removal in a future version**
_ZipError is obsolete and is no longer thrown. Use ZipException instead._

(One might argue that *is no longer thrown* hints at different behavior in past releases, but I still think it helps to clarify that code running on the current release need not worry about catching this error)

LanceAndersen replied:
> > The javadoc should be focused on the current JDK release, not the prior history of this Class
> 
> Makes sense, let's focus on the current release.
> 
> > ```
> >  * @deprecated ZipError is deprecated and subject to removal in a
> >  *  future release. This class is obsolete. Use  {@link ZipException} instead.
> > ```
> 
> The leading sentence here seems redundant, since it mostly just repeats the heading generated by javadoc:
> 
> **Deprecated, for removal: This API element is subject to removal in a future version** _ZipError is deprecated and subject to removal in a future release. This Error is obsolete and no longer thrown. Use ZipException instead._

True, I kept it as it aligns with what was done for SecurityManager and is sightly more specific. as to the element being subject to removal.

That being said,  I am OK with dropping it if we get approval in the CSR.

> 
> What do you think of dropping that first sentence, such that the rendered result would be:
> 
> **Deprecated, for removal: This API element is subject to removal in a future version** _ZipError is obsolete and is no longer thrown. Use ZipException instead._
> 
> (One might argue that _is no longer thrown_ hints at different behavior in past releases, but I still think it helps to clarify that code running on the current release need not worry about catching this error)

As I mentioned above, I am OK, with dropping it

Other options could be:

_ZipError is obsolete and is no longer used.  ZipException should be used instead_

or

_ZipError is no longer used and is obsolete. ZipException should be used instead_

eirbjo replied:
> Other options could be:
> 
> _ZipError is obsolete and is no longer used. ZipException should be used instead_

This looks good to me! I've pushed this with a an empty line above to visually separate the class description from the javadoc tags.

LanceAndersen reviewed (approved):
I think the revised proposal looks good and once you update the CSR, let me know and I will review it.

eirbjo commented:
Thanks @LanceAndersen and @liach for your patient reviews.

I have updated the Specification section of the CSR and moved it to the Proposed state for an initial round of reviews:

https://bugs.openjdk.org/browse/JDK-8338663

liach commented:
@eirbjo Can you answer CSR review question about whether to just deprecate the constructor of `ZipError` for removal at the CSR?

eirbjo commented:
> @eirbjo Can you answer CSR review question about whether to just deprecate the constructor of `ZipError` for removal at the CSR?

Thanks for prodding! I have provided an answer to this question in the CSR: https://bugs.openjdk.org/browse/JDK-8338663#comment-14709333

ExE-Boss commented:
The issue with removing `ZipError` (or any other exception class, such as `SecurityException`), is that existing classes which reference it in the `exception_table` of a `Code` attribute will fail verification.

--------------------------------------------------------------------------------

This is different from referencing removed classes in method or field descriptors, those get treated as if they had the following declaration for the purposes of verification when determining whether `checkcast`s are necessary:
```java
public abstract final class <name> extends Object {
}
```

AlanBateman commented on src/java.base/share/classes/java/util/zip/ZipError.java:
```diff
@@ -28,9 +28,12 @@
 /**
  * Signals that an unrecoverable error has occurred.
  *
+ * @deprecated ZipError is no longer used and is obsolete.
```
I agree ZipError should be deprecated. Can the JBS issue and PR title be updated as they still have "for removal".

liach replied:
Isn't this deprecation intended for removal? Or do you mean this `@deprecated` tag should note that this class will be removed in a future release?

eirbjo replied:
Alan,

This is probably better discussed in the CSR. Please see the latest comments there.

I think we're still waiting for concensus/conclusion in the CSR on whether the deprecation of the class itself should be for removal or not.

Joe suggested that the class be marked deprecated, not for removal, and that the constructor be marked deprecated, for removal. He asked for input on this suggestion, which unfortunately took me a while. 

The latest comment from me and @liach made a case for deprecating the class for removal. We are waiting for input from Joe on this.

Once the CSR concludes on a solution, I'll of course update the CSR and PR accordingly.

Thanks.

liach commented:
@ExE-Boss Indeed, the removal is intended to be binary incompatible. Just like the example in [the fucntional removal of SecurityManager(https://openjdk.org/jeps/486), users can instrument or statically preprocess their `java/util/zip/ZipError` class entries to `java/lang/InternalError` similar to the preprocessing of `System::exit` calls.

eirbjo commented:
The CSR has been approved to mark `java.util.zip.ZipError` deprecated, for removal with spec changes identical to the current state of this PR.

I plan to integrate this early next week. Just in case there are still any lingering concerns. 

Thanks for all the patient reviews and comments, here and in the CSR!

eirbjo commented:
As promised in the CSR, this change will be accompanied by a release note. A release note was drafted a while back, but I don't think anyone provided feedback.

Before integrating this change, I would love if someone could take a look at the release note and provide feedback. This way all aspects of this task can be completed:

https://bugs.openjdk.org/browse/JDK-8336843

eirbjo commented:
/integrate
